Add `Span` equality (`==` and `!=`) operators.

Ivorforce opened the pull request:
This happens to allow us to deduplicate a bunch of `operator==` from `String` and `Vector`.
It might be possible to remove explicit overloads in those functions, instead relying on implicit conversions to `Span`, but this is a bit risky compilation wise and thus out of scope for this PR. Will revisit this in a future PR.

`Span<char32_t>` comparison to `Span<char>` is somewhat questionable from a `String` perspective, because it implicitly interprets the `Span<char>` as `latin1`, although it's usually `ascii` (still correct) or utf-8 (incorrect). 
Still, from a `Span` perspective this is entirely reasonable, because individual elements report equality. It is up to callers to differentiate encodings if they matter.

## Benchmark

This PR happens to optimize the equality checks of `Packed*Array` (fundamentally typed `Vector`). The reason is that the old implementation lacked a `memcmp` optimization. I naively benchmarked a ~4x improvement on my computer. The improvement is likely to be better on modern `x86_64` machines, due to wider SIMD support (up to 16x).

I benchmarked this code:
```gdscript
extends Node2D

func _ready() -> void:
	var a := PackedInt32Array()
	a.resize(500000000)
	var b := PackedInt32Array()
	b.resize(500000000)

	for i in 20:
		a == b
```

This resulted in:
```
 ❯ hyperfine -m5 -iw1 "bin/godot.macos.editor.arm64.master --path /Users/lukas/new-game-project --quit" "bin/godot.macos.editor.arm64 --path /Users/lukas/new-game-project --quit"
Benchmark 1: bin/godot.macos.editor.arm64.master --path /Users/lukas/new-game-project --quit
  Time (mean ± σ):     11.970 s ±  0.092 s    [User: 10.632 s, System: 0.543 s]
  Range (min … max):   11.856 s … 12.081 s    5 runs

Benchmark 2: bin/godot.macos.editor.arm64 --path /Users/lukas/new-game-project --quit
  Time (mean ± σ):      3.168 s ±  0.035 s    [User: 1.863 s, System: 0.512 s]
  Range (min … max):    3.129 s …  3.221 s    5 runs

Summary
  bin/godot.macos.editor.arm64 --path /Users/lukas/new-game-project --quit ran
    3.78 ± 0.05 times faster than bin/godot.macos.editor.arm64.master --path /Users/lukas/new-game-project --quit
```

Ivorforce commented on core/templates/span.h:
```diff
@@ -113,3 +113,43 @@ constexpr uint64_t Span<T>::count(const T &p_val) const {
+		// is_trivially_equality_comparable would help, but it doesn't exist.
+		return memcmp(p_lhs.ptr(), p_rhs.ptr(), p_lhs.size() * sizeof(LHS)) == 0;
+	} else if constexpr (std::is_fundamental_v<LHS> && std::is_fundamental_v<RHS> && (std::is_signed_v<LHS> != std::is_signed_v<RHS>)) {
+		// Special case: Comparing a signed and an unsigned type.
+		// This is undefined behavior, so we need to cast to a common type before comparison.
+		using CommonType = std::common_type_t<LHS, RHS>;
+
+		// Normal case: Need to iterate the array manually.
+		for (size_t j = 0; j < p_lhs.size(); j++) {
+			if (static_cast<CommonType>(p_lhs.ptr()[j]) != static_cast<CommonType>(p_rhs.ptr()[j])) {
+				return false;
+			}
+		}
+
+		return true;
```
This is arguably a bad idea, because I'm actively circumventing our sanity check of signed comparisons.
The alternative would be to have callers handle this case. This would work for String comparisons (`char` could be cast to `uint8_t`). However, it doesn't work for non bitcastable types, e.g. `Span<int>` and `Span<uint8_t>` could not be compared with this implementation.

kiroxas reviewed (approved):
Make sense to me and would avoid to have to construct strings to compare char* against spans ([useful for other PRS](https://github.com/godotengine/godot/pull/104929#discussion_r2029025376) )

Ivorforce committed:
Add `Span` equality (`==` and `!=`) operators.
Exchange duplicate equality iteration implementations across `Vector` and `String` with the `Span` version, for a speed boost.

lawnjelly reviewed (approved):
Looks sensible, the more we can use span for these comparisons the better imo.
